File: module4/treeLevels.js
/*
tree levels

Write a function, treeLevels, that takes in the root of a binary tree. The function should return a 2-Dimensional array where each subarray represents a level of the tree.
test_00:

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f

treeLevels(a); // ->
// [
//   ['a'],
//   ['b', 'c'],
//   ['d', 'e', 'f']
// ]

test_01:

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");
const g = new Node("g");
const h = new Node("h");
const i = new Node("i");

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;
e.left = g;
e.right = h;
f.left = i;

//         a
//      /    \
//     b      c
//   /  \      \
//  d    e      f
//      / \    /
//     g  h   i

treeLevels(a); // ->
// [
//   ['a'],
//   ['b', 'c'],
//   ['d', 'e', 'f'],
//   ['g', 'h', 'i']
// ]
*/

/*
APPROACH
- we take in a binary tree and we want to return a 2D array where every subarray represents a level of the binary tree

- in this example we can use a stack to keep track of the nodes and levels
- we want to add in both the current node and the level [ node: 3, level: 0 ]
  - we tack on the level when we do the current.left and current.right operations

*/

// class Node {
//   constructor(val) {
//     this.val = val;
//     this.left = null;
//     this.right = null;
//   }
// }

// DFS recursive
const treeLevels = (root) => {
  const levels = [];
  fillLevels(root, levels, 0);
  return levels;
};

const fillLevels = (root, levels, levelNum) => {
  if (root === null) return;

  if (levels.length === levelNum) {
    levels.push([root.val]);
  } else {
    levels[levelNum].push(root.val);
  }

  fillLevels(root.left, levels, levelNum + 1);
  fillLevels(root.right, levels, levelNum + 1);
};

// DFS iterative
const treeLevels = (root) => {
  if (root === null) return [];

  let stack = [{ node: root, levelNum: 0 }];
  let levels = [];

  while (stack.length) {
    let { node, levelNum } = stack.pop();
    if (levels.length === levelNum) {
      levels.push([node.val]);
    } else {
      levels[levelNum].push(node.val);
    }

    if (node.right) stack.push({ node: node.right, levelNum: levelNum + 1 });
    if (node.left) stack.push({ node: node.left, levelNum: levelNum + 1 });
  }

  return levels;
};

// BFS
const treeLevels = (root) => {
  if (root === null) return [];

  let queue = [{ node: root, levelNum: 0 }];
  let levels = [];

  while (queue.length) {
    let { node, levelNum } = queue.pop();

    if (levels.length === levelNum) {
      levels.push([node.val]);
    } else {
      levels[levelNum].push(node.val);
    }

    if (node.left) queue.unshift({ node: node.left, levelNum: levelNum + 1 });
    if (node.right) queue.unshift({ node: node.right, levelNum: levelNum + 1 });
  }

  return levels;
};

File: module4/tree_levels.py
'''
tree levels

Write a function, tree_levels, that takes in the root of a binary tree. The function should return a 2-Dimensional list where each sublist represents a level of the tree.
test_00:

a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#      a
#    /   \
#   b     c
#  / \     \
# d   e     f

tree_levels(a) # ->
# [
#   ['a'],
#   ['b', 'c'],
#   ['d', 'e', 'f']
# ]

test_01:

a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')
g = Node('g')
h = Node('h')
i = Node('i')

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f
e.left = g
e.right = h
f.left = i

#         a
#      /    \
#     b      c
#   /  \      \
#  d    e      f
#      / \    /
#     g  h   i

tree_levels(a) # ->
# [
#   ['a'],
#   ['b', 'c'],
#   ['d', 'e', 'f'],
#   ['g', 'h', 'i']
# ]
'''

# class Node:
#   def __init__(self, val):
#     self.val = val
#     self.left = None
#     self.right = None

# DFS iterative
def tree_levels(root):
  if root is None:
    return []

  levels = []
  stack = [(root, 0)]

  while stack:
    node, level_num = stack.pop()

    if len(levels) == level_num:
      levels.append([ node.val ])
    else:
      levels[level_num].append(node.val)

    if node.right:
      stack.append((node.right, level_num + 1))
    if node.left:
      stack.append((node.left, level_num + 1))

  return levels

# DFS recursive
def tree_levels(root):
  levels = []
  _tree_levels(root, levels, 0)
  return levels

def _tree_levels(root, levels, level_num):
  if root is None:
    return;

  if len(levels) == level_num:
    levels.append([ root.val ])
  else:
    levels[level_num].append(root.val)

  _tree_levels(root.left, levels, level_num + 1)
  _tree_levels(root.right, levels, level_num + 1)

# BFS
from collections import deque

def tree_levels(root):
  if root is None:
    return []

  levels = []
  queue = deque([ (root, 0) ])

  while queue:
    node, level_num = queue.popleft()

    if len(levels) == level_num:
      levels.append([node.val])
    else:
      levels[level_num].append(node.val)

    if node.left:
      queue.append((node.left, level_num + 1))
    if node.right:
      queue.append((node.right, level_num + 1))

  return levels
